larq/quantizers.py

```diff
@@ -130,6 +130,94 @@ def approx_sign(x):
     return _binarize_with_weighted_grad(x)
 
 
+@utils.register_keras_custom_object
+class SteTern:
+    r"""
+    Ternarization function.
+    \\[
+    q(x) = \begin{cases}
+    +1 & x > \Delta \\\
+    0 & |x| > \Delta \\\
+    -1 & x < \Delta
+    \end{cases}
+    \\]
+
+    where $\Delta$ is defined as the threshold and can be passed as an argument,
+    or can be calculated as per the Ternary Weight Networks original paper, such that
+
+    \\[
+    \Delta = \frac{0.7}{n} \sum_{i=1}^{n} |W_i|
+    \\]
+    where we assume that $W_i$ is generated from a normal distribution.
+
+    The gradient is estimated using the Straight-Through Estimator
+    (essentially the Ternarization is replaced by a clipped identity on the
+    backward pass).
+    \\[\frac{\partial q(x)}{\partial x} = \begin{cases}
+    1 & \left|x\right| \leq 1 \\\
+    0 & \left|x\right| > 1
+    \end{cases}\\]
+
+    ```plot-activation
+    quantizers.SteTern
+    ```
+
+    # Arguments
+    x: Input tensor.
+    threshold value: The value for the threshold, $\Delta$.
+    ternary_weight_networks: Boolean of whether to use the Ternary Weight Networks threshold calculation. 
+
+    # Returns
+    Ternarized tensor.
+
+    # References
+    - [Ternary Weight Networks](http://arxiv.org/abs/1605.04711)
+    """
+
+    def __init__(self, threshold_value=0.1, ternary_weight_networks=False):
+        self.threshold_value = threshold_value
+        self.ternary_weight_networks = ternary_weight_networks
+
+    def __call__(self, x):
+        return self.ste_tern(x)
+
+    def ste_tern(self, x):
+
+        x = tf.clip_by_value(x, -1, 1)
+        if self.ternary_weight_networks == True:
+            threshold = self.threshold_twn(x)
+        else:
+            threshold = self.threshold_value
+
+        @tf.custom_gradient
+        def _ternarize_with_identity_grad(x):
+            def grad(dy):
+                return dy
+
+            return (
+                tf.sign(
+                    tf.add(
+                        tf.sign(tf.add(x, threshold)), tf.sign(tf.add(x, -threshold))
+                    )
+                ),
+                grad,
+            )
+
+        return _ternarize_with_identity_grad(x)
+
+    def threshold_twn(self, x):
+        x_sum = tf.reduce_sum(tf.abs(x))
+        threshold = tf.div(x_sum, tf.cast(tf.size(x), tf.float32))
```

Comment:
We should cast it to the `dtype` of the input tensor:
```suggestion
        threshold = x_sum / tf.cast(tf.size(x), x.dtype)
```

Reply:
Implemented in latest commit